File: tree/reconstructTree/reconstructBSTInOrder.java
//dfs
public TreeNode reconstruct(int[] in) {
  if(arr.length == 0)
    return null;
  return helper(arr, 0, arr.length-1);
}
private TreeNode helper(int[] arr, int left, int right){
  if(left > right)
    return null;
  int mid = (left + right)/2;
  TreeNode root= new TreeNode(arr[mid]);
  root.left = helper(arr, left, mid-1);
  root.right = helper(arr, mid+1, right);
  return root;
}

File: tree/reconstructTree/reconstructBSTInOrder.py
# sorted array to a hight balanced tree
def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
    if nums is None or len(nums) == 0:
        return None
    return helper(nums, 0, len(nums)-1)

def helper(nums, left, right):
    if left > right:
        return None
    mid = left + (right - left)//2
    root = TreeNode(nums[mid])
    root.left = helper(nums, left, mid-1)
    root.right = helper(nums, mid+1, right)
    return root



# sorted list to a hight balanced bst
def sortedListToBST(self, head: ListNode) -> TreeNode:
    if head is None:
        return head

    return self.dfs(head)

def dfs(self, head):
    # return condition
    if head is None:
        return None
    if head.next is None:
        return TreeNode(head.val)

    slow = fast = head
    # fast = head.next.next
    # while fast and fast.next:
    while fast.next and fast.next.next:
        fast = fast.next.next
        slow = slow.next
    right = slow.next
    slow.next = None

    root = TreeNode(slow.val)
    root.left = self.dfs(head)
    root.right = self.dfs(right)

    return root
